corehq/sql_db/util.py

```diff
@@ -62,6 +62,43 @@ def run_query_across_partitioned_databases(model_class, q_expression, values=Non
             yield result
 
 
+def paginate_query_across_partitioned_databases(model_class, q_expression, annotate=None, query_size=5000):
+    """
+    Runs a query across all partitioned databases in small chunks and produces a generator
+    with the results.
+
+    Iteration logic adopted from https://djangosnippets.org/snippets/1949/
+
+    :param model_class: A Django model class
+
+    :param q_expression: An instance of django.db.models.Q representing the
+    filter to apply
+
+    :param annotate: (optional) If specified, should by a dictionary of annotated fields
+    and their calculations. The dictionary will be splatted into the `.annotate` function
+
+    :return: A generator with the results
+    """
+    db_names = get_db_aliases_for_partitioned_query()
+
+    for db_name in db_names:
+        qs = model_class.objects.using(db_name)
+        if annotate:
+            qs = qs.annotate(**annotate)
+
+        qs = qs.filter(q_expression)
+        sort_col = 'pk'
+        value = 0
+        last_value = qs.order_by('-{}'.format(sort_col)).values_list(sort_col, flat=True).first()
```

Comment:
I believe the idea is so that it will not include data added after the queries start. That way it will theoretically return the same results as a single query would have (and could be a substantial amount of new data if it is long running and there are frequent writes)